vita3k/ctrl/src/udp_client.cpp

```diff
@@ -0,0 +1,315 @@
+        if (!last_pad_data)
+            last_pad_data.reset(new Response::PadData{});
+        *last_pad_data = data;
+    }
+}
+
+bool UDPClient::SetAddress(const std::string &address) {
+    std::stringstream address_ss(address);
+    std::string token;
+    std::getline(address_ss, token, ':');
+    std::string udp_input_address = token;
+    std::getline(address_ss, token, ':');
+    char *temp;
+    const std::uint16_t udp_input_port = static_cast<std::uint16_t>(std::strtol(token.c_str(), &temp, 0));
+    if (*temp != '\0') {
+        LOG_ERROR("Port number is not valid {}", token);
+        return false;
+    }
+    Reset();
+    StartCommunication(0, udp_input_address, udp_input_port);
+    return true;
+}
+
+void UDPClient::StartCommunication(std::size_t client, const std::string &host, std::uint16_t port) {
+    SocketCallback callback{ [this](Response::Version version) { OnVersion(version); },
+        [this](Response::PortInfo info) { OnPortInfo(info); },
+        [this, client](Response::PadData data) { OnPadData(data, client); } };
+    LOG_INFO("Starting communication with UDP input server on {}:{}", host, port);
+    std::lock_guard<std::mutex> guard_clients(clients_mutex);
+    clients[client].host = host;
+    clients[client].port = port;
+    clients[client].active = 0;
+    clients[client].socket = std::make_unique<Socket>(host, port, callback);
+    clients[client].thread = std::thread{ SocketLoop, clients[client].socket.get() };
+}
+
+void UDPClient::Reset() {
+    std::lock_guard<std::mutex> guard_clients(clients_mutex);
+    for (auto &client : clients) {
+        if (client.thread.joinable()) {
+            client.active = -1;
+            client.socket->Stop();
+            client.thread.join();
+        }
+    }
+    std::lock_guard<std::mutex> guard_pads(pads_mutex);
+    for (auto &pad : pads) {
+        pad.connected = false;
+    }
+}
+
+bool UDPClient::IsConnected() {
+    std::lock_guard<std::mutex> guard_pads(pads_mutex);
+    for (auto &pad : pads) {
+        if (pad.connected) {
+            if (std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - pad.last_update).count() < 5000)
+                return true;
+            else
+                pad.connected = false;
+        }
+    }
+    return false;
+}
+
+void UDPClient::SetGyroScale(float f) {
+    gyro_scale = f;
+}
+
+void UDPClient::GetGyroAccel(Util::Vec3f &gyro, uint64_t &gyro_timestamp, Util::Vec3f &accel, uint64_t &accel_timestamp) {
+    std::lock_guard<std::mutex> guard(last_pad_data_mutex);
+    if (!last_pad_data)
+        return;
+    LOG_TRACE("UDP PadData: gyro.pitch={}, gyro.roll={}, gyro.yaw={}, accel.x={}, accel.y={}, accel.z={}, motion_timestamp={}",
+        last_pad_data->gyro.pitch,
+        last_pad_data->gyro.roll,
+        last_pad_data->gyro.yaw,
+        last_pad_data->accel.x,
+        last_pad_data->accel.y,
+        last_pad_data->accel.z,
+        last_pad_data->motion_timestamp);
```

Comment:
Don't log something that will happen multiple times every second

Reply:
Even if it's a `LOG_TRACE` ? I thought it was supposed to be for this kind of logging since it would only show on the lowest log level. If that's the case then the `LOG_TRACE("PadData packet received");` from the `OnPadData()` method is actually the one that will happen even more times than this one when looking at my logs.
Should I remove both ?

Reply:
Yes please, even if it's the trace level, if a log is called more than once every second it will just bloat the log file and hide other useful logs.

Reply:
Understood, thank you. I'll remove them.